sagepay_payment.module

```diff
@@ -53,46 +56,66 @@ function sagepay_payment_check_signature($vendor_name, $security_key, $data) {
     ->execute()
     ->fetchAssoc();
-  $payment = entity_load('payment', array($info['pid']))[$info['pid']];
+  if (!$info) {
+    watchdog('sagepay_payment', 'Received notification with unknown VPSTxId.', [], WATCHDOG_ERROR);
+    return NotificationStatus::invalid('Unknown VPSTxId.');
+  }
+  $payment = entity_load_single('payment', $info['pid']);
+  if (!$payment) {
+    watchdog('sagepay_payment', 'Could not load payment for VPSTxId.', [], WATCHDOG_ERROR);
+    return NotificationStatus::error('Payment data not found.');
+  }
+  $vendor = $payment->method->controller_data['vendorname'];
+  if (!sagepay_payment_check_signature($vendor, $info['securitykey'], $_POST)) {
+    watchdog('sagepay_payment', 'Invalid signature.', [], WATCHDOG_ERROR);
+    return NotificationStatus::invalid('Invalid signature.');
+  }
 
   switch ($_POST['Status']) {
-    case 'OK': // if the transaction was authorised.
-    case 'PENDING': // (for European Payment Types only), if the transaction
-      // ...has yet to be accepted or rejected.
-      if (sagepay_payment_check_signature(
-          $payment->method->controller_data['vendorname'],
-          $info['securitykey'],
-          $_POST)) {
-        $payment->setStatus(new \PaymentStatusItem(PAYMENT_STATUS_SUCCESS));
-        $status = 'OK';
-      } else {
-        $payment->setStatus(new \PaymentStatusItem(PAYMENT_STATUS_FAILED));
-        $status = 'ERROR';
-      }
-      break;
-    case 'NOTAUTHED': // if the authorisation was failed by the bank.
-    case 'REJECTED': // if your fraud screening rules were not met.
-    case 'ABORT': // if the user decided to cancel the transaction whilst
-      // ...on our payment pages.
-    case 'ERROR': // if an error has occurred at Sage Pay. These are very
-      // infrequent, but your site should handle them anyway. They normally
-      // indicate a problem with bank connectivity.
+    case 'OK':
+      // The transaction was authorised.
+    case 'PENDING':
+      // The transaction has yet to be accepted or rejected.
+      // (European Payment Types only)
+      $payment->setStatus(new \PaymentStatusItem(PAYMENT_STATUS_SUCCESS));
+      entity_save('payment', $payment);
+      return NotificationStatus::ok($payment, 'Success');
+
+    case 'NOTAUTHED':
+      // The authorisation was failed by the bank.
+    case 'REJECTED':
+      // Fraud screening rules were not met.
+    case 'ABORT':
+      // The user decided to cancel the transaction whilst on the payment pages.
+    case 'ERROR':
+      // An error occurred on the SagePay side.
     default:
       watchdog('sagepay_payment', 'Sagepay-error: @status -> @detail', array('@status' => $_POST['Status'],  '@detail' => $_POST['StatusDetail']), WATCHDOG_ERROR);
       $payment->setStatus(new \PaymentStatusItem(PAYMENT_STATUS_FAILED));
-      $status = 'ERROR';
-      break;
+      entity_save('payment', $payment);
+      return NotificationStatus::ok($payment, 'Failed');
```

Comment:
Shouldn't this be `NotificationStatus::error($payment, 'Failed')`?

Reply:
No, that’s on purpose. The previous thing was actually a bug.

Reply:
> You can reply with a Status of either OK, INVALID or ERROR.
>
> ERROR should be used very rarely, and should ONLY be sent if something unforeseen has happened on your server or database (if you receive a notification POST for a transaction you cannot find, for instance).
>
> A Status of INVALID should be sent if you are not happy with the contents of the POST, either because the MD5 hash signatures did not match or you do not wish to proceed with the order.
>
> OK should be sent if you are happy with the notification and wish to proceed to charge the customer.

Reply:
good to know... :)